### PR TITLE
Add support for TM1638 Led and Key component

### PR DESCRIPTION
**Pull request by skykingjwc**
# What does this implement/fix?

Adds a component for the TM1638 Led and Key PCB

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**
[https://github.com/esphome/esphome-docs/pull/1991](https://github.com/esphome/esphome-docs/pull/1991)

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

display:
    platform: tm1638
    id: tm1638_display
    stb_pin: 5
    clk_pin: 18
    dio_pin: 23
    intensity: 5
    update_interval: 5s
    lambda: |-
      it.print("0123");

switch:
   - platform: tm1638
     id: MyLed
     led: 0

#Pressing the left most button will turn on the left most LED.
binary_sensor:
  - platform: tm1638
    name: "My Button"
    id: MyButton
    key: 0
    filters:
      - delayed_on: 10ms
    on_press:
     then:
        - switch.turn_on: MyLed
    on_relase
     then:
        - switch.turn_off: MyLed


```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by skykingjwc** [comment]
This isn't for the TM1637.  Not sure why that label was added.

File: esphome/components/tm1638/tm1638.cpp [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    if (*str == '.') {
+      if (pos != start_pos)
+        pos--;
+      this->buffer_[pos] |= 0b10000000;  // handles the decimal point or period
+    } else {
+      if (pos >= 8) {
+        ESP_LOGI(TAG, "TM1638 String is too long for the display!");
+        break;
+      }
+      this->buffer_[pos] = data;
```
Comment (ssieb) [inline esphome/components/tm1638/tm1638.cpp:180]: There's a valid compile warning at this line.  `data` won't be initialized if the character is outside of the range in line 175.

File: esphome/components/tm1638/tm1638.cpp [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    uint8_t data;
+
+    if (*str >= ' ' && *str <= '~') {
+      data = progmem_read_byte(&TM1638Translation::SevenSeg[*str - 32]);  // subract 32 to account for ASCII offset
+    }
+
+    if (*str == '.') {
+      if (pos != start_pos)
+        pos--;
+      this->buffer_[pos] |= 0b10000000;  // handles the decimal point or period
```
Comment (ssieb) [inline esphome/components/tm1638/tm1638.cpp]: What happens if there is more than one `.` in a row?  You will collapse them all into a single dot.

**Review by ssieb — CHANGES_REQUESTED** [review]
Why is the `tm1637` directory included in this PR?
I've tested this and it works.

File: esphome/components/tm1638/tm1638.cpp (hunk 2 above)
Comment (skykingjwc) [reply esphome/components/tm1638/tm1638.cpp]: Just pushed a fix.
https://github.com/esphome/esphome/pull/3340/commits/a91bed1c140802ee598e1c44ed7bd5b388405f7d

File: esphome/components/tm1638/tm1638.cpp (hunk 1 above)
Comment (skykingjwc) [reply esphome/components/tm1638/tm1638.cpp:180]: I agree that this needs to be addressed, but I am not seeing a compiler warning when I build using the command
`esphome run ..\tests\TM1638test.yaml  --device COM3`

TM1638test.yaml is the config file I'm using for testing.

How are you getting the warning to display?  I'll want to reproduce it so I can verify that it is fixed.

Comment (ssieb) [reply esphome/components/tm1638/tm1638.cpp:180]: I just did `esphome compile` (or `run`):
```
Compiling .pioenvs/switchtest/src/esphome/components/tm1638/tm1638.cpp.o
src/esphome/components/tm1638/tm1638.cpp: In member function 'uint8_t esphome::tm1638::TM1638Component::print(uint8_t, const char*)':
src/esphome/components/tm1638/tm1638.cpp:198:26: warning: 'data' may be used uninitialized in this function [-Wmaybe-uninitialized]
  198 |       this->buffer_[pos] = data;
      |       ~~~~~~~~~~~~~~~~~~~^~~~~~
```

Comment (skykingjwc) [reply esphome/components/tm1638/tm1638.cpp:180]: Got it.  Fix has been pushed to the repo.
https://github.com/esphome/esphome/pull/3340/commits/cca3408856e91266842173000788c21569400750

Comment (skykingjwc) [reply esphome/components/tm1638/tm1638.cpp:180]: > Hey there @glmnet, mind taking a look at this pull request as it has been labeled with an integration (`tm1637`) you are listed as a [code owner](https://github.com/esphome/esphome/blob/dev/CODEOWNERS#L204) for? Thanks! (message by CodeOwnersMention)

How can we get the TM1637 tag removed?  The TM1637 has nothing to do with this PR.

Comment (skykingjwc) [reply esphome/components/tm1638/tm1638.cpp:180]: I merged from dev, and it seems to have resolved the TM1637 issue

**Comment by skykingjwc** [comment]
Addressed all clang issues, linked documentation PR, added tests for new component in tests folder.

**Review by ssieb — APPROVED** [review]

File: esphome/components/tm1638/output/tm1638_led.cpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+namespace esphome {
+namespace tm1638 {
+
+static const char *const TAG = "tm1638.led";
+
+void TM1638Led::write_state(bool state) { tm1638_->set_led(led_, state); }
+
+void TM1638Led::dump_config() {
+  //LOG_SWITCH("", "TM1638 LED", this);
+  //ASK SSIEB about this
```
Comment (jesserockz) [inline esphome/components/tm1638/output/tm1638_led.cpp]:
```suggestion
  LOG_BINARY_OUTPUT("", "TM1638 LED", this);
```

Comment (skykingjwc) [reply esphome/components/tm1638/output/tm1638_led.cpp]: Removed comments

Comment (jesserockz) [reply esphome/components/tm1638/output/tm1638_led.cpp]: You missed the point of my suggestion, it was to change to LOG_BINARY_OUTPUT, not just remove the comments...

File: tests/tm1638test.yaml [hunk 4]
```diff
@@ -0,0 +1,165 @@
+esphome:
```
Comment (jesserockz) [inline tests/tm1638test.yaml]: You need to add to an existing test file (1,2,3,4,5), new files are ignored

File: esphome/components/tm1638/output/tm1638_led.cpp (hunk 3 above)
Comment (skykingjwc) [reply esphome/components/tm1638/output/tm1638_led.cpp]: My bad,  must have been getting tired. 
Added LOG_BINARY_OUTPUT
c0f9d8b

File: tests/tm1638test.yaml (hunk 4 above)
Comment (skykingjwc) [reply tests/tm1638test.yaml]: Added tests to test5.yaml

Comment (jesserockz) [reply tests/tm1638test.yaml]: Dont forget to delete this yaml file too

File: esphome/components/tm1638/switch/__init__.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import switch
+from esphome.const import CONF_ID, CONF_LED
+from ..display import tm1638_ns, TM1638Component, CONF_TM1638_ID
+
+TM1638SwitchLed = tm1638_ns.class_("TM1638SwitchLed", switch.Switch)
```
Comment (jesserockz) [inline esphome/components/tm1638/switch/__init__.py]:
```suggestion
TM1638SwitchLed = tm1638_ns.class_("TM1638SwitchLed", switch.Switch, cg.Component)
```

File: esphome/components/tm1638/switch/__init__.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+
+CONFIG_SCHEMA = switch.SWITCH_SCHEMA.extend(
+    {
+        cv.GenerateID(): cv.declare_id(TM1638SwitchLed),
+        cv.GenerateID(CONF_TM1638_ID): cv.use_id(TM1638Component),
+        cv.Required(CONF_LED): cv.int_range(min=0, max=7),
+    }
+)
```
Comment (jesserockz) [inline esphome/components/tm1638/switch/__init__.py]:
```suggestion
).extend(cv.COMPONENT_SCHEMA)
```

File: esphome/components/tm1638/switch/__init__.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await switch.register_switch(var, config)
```
Comment (jesserockz) [inline esphome/components/tm1638/switch/__init__.py]:
```suggestion
    var = await switch.new_switch(config)
    await cg.register_component(var, config)
```

File: esphome/components/tm1638/output/__init__.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import output
+from esphome.const import CONF_ID, CONF_LED
+from ..display import tm1638_ns, TM1638Component, CONF_TM1638_ID
+
+TM1638OutputLed = tm1638_ns.class_("TM1638OutputLed", output.BinaryOutput)
```
Comment (jesserockz) [inline esphome/components/tm1638/output/__init__.py]:
```suggestion
TM1638OutputLed = tm1638_ns.class_("TM1638OutputLed", output.BinaryOutput, cg.Component)
```

File: esphome/components/tm1638/output/__init__.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+
+CONFIG_SCHEMA = output.BINARY_OUTPUT_SCHEMA.extend(
+    {
+        cv.GenerateID(): cv.declare_id(TM1638OutputLed),
+        cv.GenerateID(CONF_TM1638_ID): cv.use_id(TM1638Component),
+        cv.Required(CONF_LED): cv.int_range(min=0, max=7),
+    }
+)
```
Comment (jesserockz) [inline esphome/components/tm1638/output/__init__.py]:
```suggestion
).extend(cv.COMPONENT_SCHEMA)
```

File: esphome/components/tm1638/output/__init__.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await output.register_output(var, config)
```
Comment (jesserockz) [inline esphome/components/tm1638/output/__init__.py:21]:
```suggestion
    await output.register_output(var, config)
    await cg.register_component(var, config)
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: tests/tm1638test.yaml (hunk 4 above)
Comment (skykingjwc) [reply tests/tm1638test.yaml]: Deleted tm1638test.yaml from repo

File: esphome/components/tm1638/binary_sensor/__init__.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        cv.GenerateID(): cv.declare_id(TM1638Key),
+        cv.GenerateID(CONF_TM1638_ID): cv.use_id(TM1638Component),
+        cv.Required(CONF_KEY): cv.int_range(min=0, max=15),
+    }
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await binary_sensor.register_binary_sensor(var, config)
```
Comment (ssieb) [inline esphome/components/tm1638/binary_sensor/__init__.py]:
```suggestion
    var = await binary_sensor.new_binary_sensor(config)
```

File: esphome/components/tm1638/tm1638.cpp [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+float TM1638Component::get_setup_priority() const { return setup_priority::PROCESSOR; }
+
+void TM1638Component::display() {
+  for (uint8_t i = 0; i < 8; i++) {
+    this->set_7seg_(i, buffer_[i]);
+  }
+}
+
+void TM1638Component::reset_(bool on_off) {
```
Comment (ssieb) [inline esphome/components/tm1638/tm1638.cpp]: I don't see anywhere that this is called with `true`, so you can remove the parameter.

File: tests/tm1638test.yaml (hunk 4 above)
Comment (ssieb) [reply tests/tm1638test.yaml]: It's still here.  You need to `git rm` it.

**Review by ssieb — CHANGES_REQUESTED** [review]

File: esphome/components/tm1638/tm1638.cpp (hunk 12 above)
Comment (skykingjwc) [reply esphome/components/tm1638/tm1638.cpp]: Removed unnecessary parameter

**Review by ssieb — APPROVED** [review]

**Comment by ssieb** [comment]
I will commit after verifying something with the contributor.

**Pull request merged**